### ISSUE TITLE
`debug_print` renders unsigned vector values as signed integers

### ISSUE THREAD

**Issue by yixiaoer** [jax-ml/jax#40860]
### Description

On TPU v4, `pl.debug_print` misprints unsigned values as signed integers. The kernel prints and copies the same array; its returned values are correct.

## Environment

- Python 3.12
- JAX/jaxlib 0.11.2
- libtpu 0.0.48

## Reproducer

Save as `repro_unsigned_print.py` and run `python repro_unsigned_print.py`. Logging is enabled before importing JAX.

```python
import os

os.environ["JAX_PLATFORMS"] = "tpu"
os.environ["TPU_STDERR_LOG_LEVEL"] = "0"
os.environ["JAX_ENABLE_COMPILATION_CACHE"] = "false"

import jax
import numpy as np
from jax.experimental import pallas as pl

host = np.tile(np.array([0, 2**31 - 1, 2**31, 2**32 - 1], dtype=np.uint32), (3, 1))

def kernel(x_ref, out_ref):
    value = x_ref[...]
    pl.debug_print("unsigned_u32 {}", value)
    out_ref[...] = value

run = jax.jit(
    pl.pallas_call(kernel, out_shape=jax.ShapeDtypeStruct(host.shape, host.dtype)),
    compiler_options={"xla_tpu_enable_log_recorder": "true"},
)
result = run(jax.device_put(host, jax.devices("tpu")[0]))
result.block_until_ready()
jax.effects_barrier()

actual = np.asarray(result)
np.testing.assert_array_equal(actual, host)
print("Returned dtype:", actual.dtype, flush=True)
print("Returned first row:", actual[0].tolist(), flush=True)
```

## Expected versus actual

Each array has three identical rows. Comparing the first row:

| Observation | Type | Values |
|---|---|---|
| Input | `uint32` | `[0, 2147483647, 2147483648, 4294967295]` |
| Returned array (correct) | `uint32` | `[0, 2147483647, 2147483648, 4294967295]` |
| `pl.debug_print` log (incorrect) | `s32` | `[0, 2147483647, -2147483648, -1]` |

The log should preserve unsigned values and type labels. Instead, valid uint32 values such as `4294967295` are displayed as signed integers (`-1`).

The assertion checks the returned array, not the log, so it passes despite the incorrect text.

Separate comparison on the same stack also reproduced:

| Input dtype and values | Logged values | Logged type |
|---|---|---|
| uint8: `128, 255` | `-128, -1` | `s8` |
| uint16: `32768, 65535` | `-32768, -1` | `s16` |

For all three dtypes, the returned arrays match the inputs; the printed log values do not.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.11.2
jaxlib: 0.11.2
numpy:  2.5.2
python: 3.12.14 (main, Aug 14 2026, 15:34:45) [Clang 22.1.3 ]
device info: TPU v4-4, 4 local devices
process_count: 1
platform: uname_result(system='Linux', node='t1v-n-1e6669f9-w-0', release='5.19.0-1022-gcp', version='#24~22.04.1-Ubuntu SMP Sun Apr 23 09:51:08 UTC 2023', machine='x86_64')
JAX_PLATFORMS=tpu
```

**Comment by a1l2v** [comment]
Hi @yixiaoer I'd like to take up this issue. Could you please assign it to me?

Looking into the Mosaic lowering code in  jax/_src/pallas/mosaic/lowering.py , a potential root cause appears to be in _debug_print_rule (#L5640), where IR types are constructed using _dtype_to_ir_type. 
In _dtype_to_ir_type (#L591-L596), integer types are converted into signless MLIR integer types , which drops the explicit unsigned metadata before TPU log generation.

Plan of Action:
1) Reproduce and trace the exact printing behavior for uint8, uint16, and uint32 dtypes under TPU runtime execution (pl.debug_print).
2) first try - preserving signedness in the lowering rule (This may not work if the Mosaic TPU compiler doesn't support unsigned element types in MemRefs)
3) if not the first approach, i would keep Signless MemRef, But Bitcast Before Logging resolving the signed output behavior
4) Add regression tests covering unsigned integer formatting in pl.debug_print
5) Submit a PR with the fix.

I'm happy to adjust the approach based on any feedback or guidance

**Comment by VanshBhardwaj24** [comment]
cc @yixiaoer @justinjfu — opened a fix for this in [#40871](https://github.com/jax-ml/jax/pull/40871). Root cause is the signless MLIR conversion in _dtype_to_ir_type dropping unsigned info before tpu.log_buffer; fix reinterpret-casts to the correct unsigned type right before printing. Would appreciate a look when you get a chance.
